pagespeed/system/controller_manager.cc

```diff
@@ -0,0 +1,329 @@
+#include "pagespeed/kernel/base/string_util.h"
+#include "pagespeed/system/system_rewrite_options.h"
+
+namespace net_instaweb {
+
+
+ControllerManager::ControllerManager(
+    SystemRewriteDriverFactory* factory,
+    ThreadSystem* thread_system,
+    FileSystem* file_system,
+    Timer* timer,
+    MessageHandler* handler) : factory_(factory),
+                               thread_system_(thread_system),
+                               file_system_(file_system),
+                               timer_(timer),
+                               handler_(handler),
+                               babysitter_pid_(-1) { }
+
+void ControllerManager::SetUpSignalHandlers() {
+  // We need to clear inherited signal handlers. There's no portable way to get
+  // a list of all possible signals, and they're not even guaranteed to be in
+  // order.  But NSIG is usually defined these days, and if it is then we just
+  // want ascending numbers up to to NSIG.
+  for (int i = 0; i < NSIG; i++) {
+    signal(i, SIG_DFL);
+  }
+}
+
+ControllerManager::ProcessDeathWatcherThread::ProcessDeathWatcherThread(
+    ThreadSystem* thread_system,
+    Timer* timer,
+    MessageHandler* handler,
+    pid_t master_pid,
+    pid_t babysitter_pid) : Thread(thread_system,
+                                   "master process death watcher",
+                                   ThreadSystem::kDetached),
+                            timer_(timer),
+                            handler_(handler),
+                            master_pid_(master_pid),
+                            babysitter_pid_(babysitter_pid) {
+}
+
+bool IsProcessRunning(pid_t pid) {
+  // "If sig is 0, then no signal is sent, but error checking is still
+  //  performed; this can be used to check for the existence of a process ID
+  //  or process group ID."
+  return kill(pid, 0) == 0;
+}
+
+void ControllerManager::ProcessDeathWatcherThread::Run() {
+  handler_->Message(
+      kInfo, "Watching master (%d) and babysitter (%d) to exit if they do.",
+      master_pid_, babysitter_pid_);
+  while (IsProcessRunning(master_pid_) &&
+         IsProcessRunning(babysitter_pid_)) {
+    timer_->SleepUs(30 * 1000 /* 0.03s */);
+  }
+
+  if (!IsProcessRunning(master_pid_)) {
+    handler_->Message(
+        kInfo, "Master process (%d) exited; controller shutting down.",
+        master_pid_);
+  } else {
+    handler_->Message(
+        kInfo, "Babysitter process (%d) exited; controller shutting down.",
+        babysitter_pid_);
+  }
+
+  // TODO(jefftk): Once the controller is doing real work and there's real
+  // cleanup work to do, we can initiate that from here.
+
+  exit(EXIT_SUCCESS);
+
+  // Because the controller is exiting via exit(), the babysitter process will
+  // see that and quit itself instead of restarting it.
+}
+
+pid_t ControllerManager::ReadPidFromFile(
+    const GoogleString& filename) {
+  if (file_system_->Exists(
+          filename.c_str(), handler_).is_true()) {
+    GoogleString pid_buffer;
```

Comment:
Why the explicit exists? (Partly asking since the file can disappear between exists and read... though of course it also can after the read)